Enable HW-accelerated implementations of SHA256 for MSVC builds

hebasto opened the pull request:
This PR enables AVX2, SSE4.1 and x86 SHA-NI hardware-accelerated implementations of SHA256 to replace the "standard" default implementation when building on Windows.

**Testing Note:** At runtime, the SHA-NI implementation is dynamically selected and available only if the CPU has the `sha_ni` flag set.

Here are the benchmark results on my machine using "Release" binaries:

- the master branch @ 47da4f9b716d11294d4fb0f30b04a7bcf128cc14:
```
> build\bin\Release\bench_bitcoin.exe -filter="SHA256.*_SHANI"

|             ns/byte |              byte/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               11.35 |       88,074,183.58 |    0.3% |      0.01 | `SHA256D64_1024_SHANI using the 'standard' SHA256 implementation`
|               13.07 |       76,522,397.95 |    0.4% |      0.01 | `SHA256_32b_SHANI using the 'standard' SHA256 implementation`
|                4.26 |      234,554,580.85 |    0.9% |      0.05 | `SHA256_SHANI using the 'standard' SHA256 implementation`

> build\bin\Release\bench_bitcoin.exe -filter="SHA256D64_1024.*"

|             ns/byte |              byte/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               11.29 |       88,586,104.35 |    0.3% |      0.01 | `SHA256D64_1024_AVX2 using the 'standard' SHA256 implementation`
|               11.36 |       88,062,348.83 |    0.9% |      0.01 | `SHA256D64_1024_SHANI using the 'standard' SHA256 implementation`
|               11.33 |       88,275,862.07 |    0.9% |      0.01 | `SHA256D64_1024_SSE4 using the 'standard' SHA256 implementation`
|               11.33 |       88,263,973.06 |    0.8% |      0.01 | `SHA256D64_1024_STANDARD using the 'standard' SHA256 implementation`
```

- this PR:

```
> build\bin\Release\bench_bitcoin.exe -filter="SHA256.*_SHANI"

|             ns/byte |              byte/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                1.27 |      790,145,684.72 |    0.5% |      0.01 | `SHA256D64_1024_SHANI using the 'x86_shani(1way;2way)' SHA256 implementation`
|                1.97 |      508,638,668.32 |    0.8% |      0.01 | `SHA256_32b_SHANI using the 'x86_shani(1way;2way)' SHA256 implementation`
|                0.63 |    1,587,301,587.30 |    0.5% |      0.01 | `SHA256_SHANI using the 'x86_shani(1way;2way)' SHA256 implementation`

> build\bin\Release\bench_bitcoin.exe -filter="SHA256D64_1024.*"

|             ns/byte |              byte/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                4.02 |      248,878,761.99 |    0.3% |      0.01 | `SHA256D64_1024_AVX2 using the 'standard;sse41(4way);avx2(8way)' SHA256 implementation`
|                1.27 |      787,950,595.69 |    0.8% |      0.01 | `SHA256D64_1024_SHANI using the 'x86_shani(1way;2way)' SHA256 implementation`
|                7.00 |      142,826,631.80 |    0.8% |      0.01 | `SHA256D64_1024_SSE4 using the 'standard;sse41(4way)' SHA256 implementation`
|               11.23 |       89,055,578.20 |    1.0% |      0.01 | `SHA256D64_1024_STANDARD using the 'standard' SHA256 implementation`
```

laanwj commented:
Concept ACK—but it does makes me wonder why AVX2 specifically? Would, say, enabling SHANI (which is better on hw that supports it) give a lot of extra work?

hebasto commented:
> Concept ACK—but it does makes me wonder why AVX2 specifically? Would, say, enabling SHANI (which is better on hw that supports it) give a lot of extra work?

I have no hardware that supports SHANI (

hebasto commented:
> > I have no hardware that supports SHANI (
> 
> That's a valid reason. I guess the base work in this PR will make it easier for people who care, to add support the other instruction sets too?

Exactly.

sipsorcery commented:
tACK 70f36d816e823d300361f872923f2c3a5f90d476.

*I'm not qualified to verify the safety of changing crypto logic.

I performed an msvc build and then ran the benchmark with and without this PR and there doesn't seem to be much of a difference for `SHA256` on my Intel i7 CPU based PC. I'm not sure if that as expected or not?

On `master`:
````
|             ns/byte |              byte/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               12.70 |       78,766,206.15 |    2.6% |      0.14 | `SHA1`
|               16.99 |       58,841,181.77 |    1.2% |      0.19 | `SHA256`
|               46.23 |       21,630,470.66 |   11.5% |      0.04 | :wavy_dash: `SHA256D64_1024` (Unstable with ~1.0 iters. Increase `minEpochIterations` to e.g. 10)
|               36.54 |       27,365,792.76 |    1.8% |      0.01 | `SHA256_32b`
|               24.93 |       40,108,453.26 |    3.8% |      0.29 | `SHA3_256_1M`
|               11.64 |       85,924,678.43 |    1.7% |      0.13 | `SHA512`
````

With #24773:
````
|             ns/byte |              byte/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               13.79 |       72,504,223.37 |    4.9% |      0.16 | `SHA1`
|               18.06 |       55,382,638.65 |    4.6% |      0.21 | `SHA256`
|               88.30 |       11,324,888.97 |   13.0% |      0.06 | :wavy_dash: `SHA256D64_1024` (Unstable with ~1.0 iters. Increase `minEpochIterations` to e.g. 10)
|               56.46 |       17,710,386.66 |   18.7% |      0.01 | :wavy_dash: `SHA256_32b` (Unstable with ~618.0 iters. Increase `minEpochIterations` to e.g. 6180)
|               28.18 |       35,489,686.70 |    2.3% |      0.33 | `SHA3_256_1M`
|               12.35 |       80,963,792.99 |    6.3% |      0.14 | :wavy_dash: `SHA512` (Unstable with ~1.0 iters. Increase `minEpochIterations` to e.g.10)
````

hebasto commented:
Rebased on top of the merged #24772.

hebasto commented:
Converted to a "draft" until resolving discussion in #24774.

fanquake commented:
Closing for now, given you've closed the remaining base PR (#24774). Let me know if this is incorrect / should be reopened.

hebasto commented:
> Would, say, enabling SHANI (which is better on hw that supports it) give a lot of extra work?

Implemented.

> Converted to a "draft" until resolving discussion in https://github.com/bitcoin/bitcoin/pull/24774.

No longer based on #24774.

fanquake commented on src/random.cpp:
```diff
@@ -67,7 +67,7 @@ static inline int64_t GetPerformanceCounter() noexcept
 }
 
-#ifdef HAVE_GETCPUID
+#if defined(HAVE_GETCPUID) && !defined(_MSC_VER)
```
If `GetCPUID` is now implemented for MSVC, why is this skipped?

hebasto replied:
> If `GetCPUID` is now implemented for MSVC

It is implemented in the first commit.

> why is this skipped?

MSVC does not support inline assembly on x64: https://github.com/bitcoin/bitcoin/blob/593979494042e77b5271d42b8f6e7b514e0007d8/src/random.cpp#L129-L136

fanquake replied:
> It is implemented in the first commit.

I'm aware. That's why I said "is now implemented".

hebasto replied:
1. This `#if ... #else ... #endif` [block](https://github.com/hebasto/bitcoin/blob/915bdbdf8c87e20426161bbf393d1036659813a2/src/random.cpp#L69-L185) guards two static functions--`GetRdRand` and `GetRdSeed`--which are not implemented for Windows. Not adding `&& !defined(_MSC_VER)` ends with an error:
```
C:\Users\hebasto\bitcoin\src\random.cpp(137,1): fatal  error C1189: #error:  "RdRand is only supported on x86 and x86_64" [C:\Users\hebasto\bitcoin\build_msvc\li
bbitcoin_util\libbitcoin_util.vcxproj]
```

2. > And I don't understand the comment about MSVC and assembly.

Windows implementation for the `GetRdRand` and `GetRdSeed` was not added because "MSVC does not support inline assembly on x64".

3. > Was `GetCPUID` not compiling already?

It was. On Windows, it is used in the `SHA256AutoDetect` and `AddCPUID` functions.

theuni replied:
Hmm, ok, we are talking around each-other. I'll ask some detailed questions in code and see if we can clear it up.

fanquake commented on src/compat/attribute.h:
```diff
@@ -0,0 +1,16 @@
+// Copyright (c) 2022 The Bitcoin Core developers
```
We already have src/attributes.h. Any reason this is another file?

hebasto replied:
Thanks!. [Updated](https://github.com/bitcoin/bitcoin/pull/24773#issuecomment-1329464328).

fanquake commented on src/compat/attribute.h:
```diff
@@ -0,0 +1,16 @@
+#define BITCOIN_COMPAT_ATTRIBUTE_H
+
+#if defined(__GNUC__)
+#define FORCED_INLINE inline __attribute__((always_inline))
```
If we are going to add this, `ALWAYS_INLINE` seems like a better name.

hebasto replied:
Thanks!. [Updated](https://github.com/bitcoin/bitcoin/pull/24773#issuecomment-1329464328).

hebasto commented:
Updated b27defef4218b249a2ec5ec0dd033f62add8daf1 -> 674d3e9a405bebd6388632dbc23b733c65c7887c ([pr24773.03](https://github.com/hebasto/bitcoin/commits/pr24773.03) -> [pr24773.04](https://github.com/hebasto/bitcoin/commits/pr24773.04), [diff](https://github.com/hebasto/bitcoin/compare/pr24773.03..pr24773.04)):

- addressed @fanquake's comments

hebasto commented:
Friendly ping @sipsorcery :)

theuni reviewed (changes requested):
Concept ACK. Coming to this after kill/shill/merge, sorry for the late re-review.

Would you mind splitting up the inline changes from the rest? Looks good to me, but as it introduces a general abstraction in the crypto code I think it'd be nice not to restrict review to win32 devs for that part.

After that the changes here look pretty uncontroversial to me (modulo actually verifying the cpuid bitwise stuff).

theuni commented:
Sorry, yes, I meant to factor out those changes into a separate PR.

hebasto commented:
> Sorry, yes, I meant to factor out those changes into a separate PR.

Thanks! Done in https://github.com/bitcoin/bitcoin/pull/27575.

hebasto commented:
> > Sorry, yes, I meant to factor out those changes into a separate PR.
> 
> Thanks! Done in #27575.

Rebased on top of the #27575 and drafted. Please review #27575 first.

hebasto commented:
I had a chance to benchmark on a Windows laptop equipped with Intel Core i9 -12950HX with the `sha_ni` flag set.

No significant difference has been observed. Going to close this PR.

hebasto commented on src/compat/cpuid.h:
```diff
@@ -23,4 +23,22 @@ void static inline GetCPUID(uint32_t leaf, uint32_t subleaf, uint32_t& a, uint32
 #endif // defined(__x86_64__) || defined(__amd64__) || defined(__i386__)
+
+#if defined(_MSC_VER) && defined(_M_X64)
+#define HAVE_GETCPUID
```
> Does MSVC not define `__x86_64__` or `__amd64__` or `__i386__` ?

No, it does not. Easy to verify with the following diff:
```diff
--- a/src/compat/cpuid.h
+++ b/src/compat/cpuid.h
@@ -6,6 +6,7 @@
 #define BITCOIN_COMPAT_CPUID_H

 #if defined(__x86_64__) || defined(__amd64__) || defined(__i386__)
+#error
 #define HAVE_GETCPUID

 #include <cpuid.h>

```

hebasto commented on src/crypto/sha256.cpp:
```diff
@@ -576,6 +576,17 @@ bool AVXEnabled()
+
+bool AVXEnabled()
+{
+    auto a = _xgetbv(_XCR_XFEATURE_ENABLED_MASK);
```
> Why does the existing asm function not work?

From https://learn.microsoft.com/en-us/cpp/assembler/inline/inline-assembler:
> Inline assembly is not supported on the ARM and x64 processors.

---

> What is the specific error? Does it not like the inline asm? Or does it just have a problem with the opcode?

With the diff as follows:
```diff
--- a/src/crypto/sha256.cpp
+++ b/src/crypto/sha256.cpp
@@ -567,7 +567,6 @@ bool SelfTest() {
     return true;
 }

-#if defined(USE_ASM) && (defined(__x86_64__) || defined(__amd64__) || defined(__i386__))
 /** Check whether the OS has enabled AVX registers. */
 bool AVXEnabled()
 {
@@ -575,7 +574,6 @@ bool AVXEnabled()
     __asm__("xgetbv" : "=a"(a), "=d"(d) : "c"(0));
     return (a & 6) == 6;
 }
-#endif
 } // namespace

```

I have the following error:
```
C:\bitcoin\src\crypto\sha256.cpp(574,5): error C3861: '__asm__': identifier not found [C:\bitcoin\build_msvc\libbitcoin_crypto\libbitcoin_crypto.vcxproj]
```

hebasto commented:
Rebased on top of the merged https://github.com/bitcoin/bitcoin/pull/27575 and undrafted.

hebasto commented:
Drafted again as it seems worth going https://github.com/bitcoin/bitcoin/pull/27598 and https://github.com/bitcoin/bitcoin/pull/27615 first.

> UPD. Well, the reason is that
> 
> https://github.com/bitcoin/bitcoin/blob/322ec63b01499c1ec52d3912ee382ebd59f2366b/src/bench/bench_bitcoin.cpp#L64

Fixed in https://github.com/bitcoin/bitcoin/pull/27615.

hebasto commented:
> Rebase for fresh CI, if still relevant?

Thanks! Done.

hebasto commented:
Rebase on top of the merged #27598 and undrafted.

sipsorcery commented:
tACK e94ae810a55da16fce6040714677b9d50781bebd.

Relevant `bench_bitcoin.exe` results from my Intel i9 CPU machine. Looks up 5 to 10x faster than my i7 results above.

```
|             ns/byte |              byte/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                1.37 |      729,927,007.30 |    0.3% |      0.02 | `SHA1`
|                2.20 |      453,902,805.03 |    3.0% |      0.01 | `SHA256D64_1024_AVX2 using the 'standard,sse41(4way),avx2(8way)' SHA256 implementation`
|                3.32 |      301,453,541.86 |    0.3% |      0.01 | `SHA256D64_1024_SHANI using the 'standard,sse41(4way)' SHA256 implementation`
|                3.32 |      300,830,846.91 |    0.3% |      0.01 | `SHA256D64_1024_SSE4 using the 'standard,sse41(4way)' SHA256 implementation`
|                9.18 |      108,890,919.66 |    3.2% |      0.01 | `SHA256D64_1024_STANDARD using the 'standard' SHA256 implementation`
|                7.42 |      134,746,633.08 |    0.4% |      0.01 | `SHA256_32b_AVX2 using the 'standard,sse41(4way),avx2(8way)' SHA256 implementation`
|                7.41 |      134,917,592.95 |    0.3% |      0.01 | `SHA256_32b_SHANI using the 'standard,sse41(4way)' SHA256 implementation`
|                8.03 |      124,525,069.29 |    6.0% |      0.01 | :wavy_dash: `SHA256_32b_SSE4 using the 'standard,sse41(4way)' SHA256 implementation` (Unstable with ~3,798.2 iters. Increase `minEpochIterations` to e.g. 37982)
|                7.46 |      133,974,122.00 |    0.7% |      0.01 | `SHA256_32b_STANDARD using the 'standard' SHA256 implementation`
|                3.55 |      281,825,099.34 |    3.2% |      0.04 | `SHA256_AVX2 using the 'standard,sse41(4way),avx2(8way)' SHA256 implementation`
|                3.45 |      289,939,112.79 |    2.4% |      0.04 | `SHA256_SHANI using the 'standard,sse41(4way)' SHA256 implementation`
|                3.47 |      287,952,084.77 |    1.1% |      0.04 | `SHA256_SSE4 using the 'standard,sse41(4way)' SHA256 implementation`
|                3.44 |      290,875,243.61 |    0.6% |      0.04 | `SHA256_STANDARD using the 'standard' SHA256 implementation`
|                3.39 |      294,828,704.52 |    1.9% |      0.04 | `SHA3_256_1M`
|                2.21 |      453,185,896.85 |    2.0% |      0.02 | `SHA512`
```

hebasto commented:
I had an opportunity to benchmark this branch on a machine with [Intel Core i9-12950HX](https://openbenchmarking.org/s/Intel+Core+i9-12950HX), which has the `sha_ni` flag set. Results are posted in the PR description.

The performance gain is huge...

cc @sipa @theuni

m3dwards commented on build_msvc/bitcoin_config.h.in:
```diff
@@ -29,6 +29,9 @@
 /* Copyright year */
 #define COPYRIGHT_YEAR $
 
+/* Define this symbol to build in assembly routines */
+#define USE_ASM
```
NIT: From what I understand we are enabling the special CPU instructions using intrinsics and not inline assembly. This makes the comment a little misleading and makes the symbol unfortunately named.

Perhaps the symbol could be?

```suggestion
/* Define this symbol to enable CPU SHA Extensions */
#define USE_SHA_EXTENTIONS
```

or?

```suggestion
/* Define this symbol to enable CPU Extensions */
#define USE_INTRINSICS
```

m3dwards reviewed (commented):
Tested e94ae81 on Windows 11 with VS2022. CPU: i7-4870HQ and saw big speed improvement.

On master:

```
$ ./src/bench_bitcoin.exe -filter=SHA256D64.*

|             ns/byte |              byte/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               11.57 |       86,459,102.90 |    0.0% |      0.01 | `SHA256D64_1024_AVX2 using the 'standard' SHA256 implementation`
|               11.89 |       84,074,406.67 |    0.0% |      0.01 | `SHA256D64_1024_SHANI using the 'standard' SHA256 implementation`
|               11.97 |       83,559,862.30 |    0.6% |      0.01 | `SHA256D64_1024_SSE4 using the 'standard' SHA256 implementation`
|               11.90 |       84,031,286.06 |    0.0% |      0.01 | `SHA256D64_1024_STANDARD using the 'standard' SHA256 implementation`
```

Using #24773

```
$ ./src/bench_bitcoin.exe -filter=SHA256D64.*

|             ns/byte |              byte/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                2.77 |      361,557,983.01 |    0.1% |      0.01 | `SHA256D64_1024_AVX2 using the 'standard,sse41(4way),avx2(8way)' SHA256 implementation`
|                4.33 |      230,699,639.18 |    0.1% |      0.01 | `SHA256D64_1024_SHANI using the 'standard,sse41(4way)' SHA256 implementation`
|                4.57 |      219,037,433.16 |    2.1% |      0.01 | `SHA256D64_1024_SSE4 using the 'standard,sse41(4way)' SHA256 implementation`
|               11.88 |       84,171,590.03 |    0.2% |      0.01 | `SHA256D64_1024_STANDARD using the 'standard' SHA256 implementation`
```

One tiny nit on naming, not important to hold it up.

hebasto commented:
Rebased.

achow101 commented:
Deferring to after cmake

sedited commented:
@hebasto This change has lingered for a long time. We should either move forward with it now, or close it again.

hebasto committed:
compat: Add MSVC implementation of `GetCPUID()`

hebasto committed:
crypto: Add MSVC implementation of `AVXEnabled()`

hebasto committed:
cmake: Enable SSE4.1 support for MSVC

hebasto committed:
cmake: Enable AVX2 support for MSVC

hebasto committed:
cmake: Enable SHA-NI support for MSVC

hebasto commented:
> Deferring to after cmake

Rebased. The PR description has been updated.

achow101 commented:
Benchmarks on a AMD 5950x:

Master (5f33da9aa30f):

```
> .\build\bin\Release\bench_bitcoin.exe -filter="SHA256.*"
```

|             ns/byte |              byte/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                7.44 |      134,363,915.94 |    0.0% |      0.01 | `SHA256D64_1024_AVX2 using the 'standard' SHA256 implementation`
|                7.56 |      132,249,016.24 |    1.3% |      0.01 | `SHA256D64_1024_SHANI using the 'standard' SHA256 implementation`
|                7.70 |      129,954,392.23 |    3.9% |      0.01 | `SHA256D64_1024_SSE4 using the 'standard' SHA256 implementation`
|                7.70 |      129,954,392.23 |    3.9% |      0.01 | `SHA256D64_1024_STANDARD using the 'standard' SHA256 implementation`
|                9.23 |      108,395,773.29 |    3.6% |      0.01 | `SHA256_32b_AVX2 using the 'standard' SHA256 implementation`
|                9.29 |      107,597,034.45 |    4.7% |      0.01 | `SHA256_32b_SHANI using the 'standard' SHA256 implementation`
|                9.13 |      109,481,765.83 |    2.6% |      0.01 | `SHA256_32b_SSE4 using the 'standard' SHA256 implementation`
|                9.24 |      108,205,791.62 |    3.8% |      0.01 | `SHA256_32b_STANDARD using the 'standard' SHA256 implementation`
|                3.91 |      255,591,054.31 |    0.1% |      0.04 | `SHA256_AVX2 using the 'standard' SHA256 implementation`
|                3.92 |      255,264,837.27 |    0.2% |      0.04 | `SHA256_SHANI using the 'standard' SHA256 implementation`
|                3.92 |      255,303,939.34 |    0.1% |      0.04 | `SHA256_SSE4 using the 'standard' SHA256 implementation`
|                3.93 |      254,381,725.22 |    0.4% |      0.04 | `SHA256_STANDARD using the 'standard' SHA256 implementation`

PR (21da1452af35):

|             ns/byte |              byte/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                9.41 |      106,311,947.44 |    0.4% |      0.01 | `SHA256D64_1024_AVX2 using the 'standard;sse41(4way);avx2(8way)' SHA256 implementation`
|                0.72 |    1,396,138,278.30 |    2.5% |      0.01 | `SHA256D64_1024_SHANI using the 'x86_shani(1way;2way)' SHA256 implementation`
|                4.82 |      207,348,660.62 |    1.4% |      0.01 | `SHA256D64_1024_SSE4 using the 'standard;sse41(4way)' SHA256 implementation`
|                7.68 |      130,147,949.56 |    3.7% |      0.01 | `SHA256D64_1024_STANDARD using the 'standard' SHA256 implementation`
|                9.20 |      108,657,750.41 |    3.4% |      0.01 | `SHA256_32b_AVX2 using the 'standard;sse41(4way);avx2(8way)' SHA256 implementation`
|                1.27 |      785,448,659.59 |    2.8% |      0.01 | `SHA256_32b_SHANI using the 'x86_shani(1way;2way)' SHA256 implementation`
|                9.16 |      109,196,025.29 |    2.8% |      0.01 | `SHA256_32b_SSE4 using the 'standard;sse41(4way)' SHA256 implementation`
|                9.19 |      108,780,236.75 |    4.0% |      0.01 | `SHA256_32b_STANDARD using the 'standard' SHA256 implementation`
|                3.92 |      255,316,976.03 |    0.2% |      0.04 | `SHA256_AVX2 using the 'standard;sse41(4way);avx2(8way)' SHA256 implementation`
|                0.44 |    2,285,975,540.06 |    0.4% |      0.01 | `SHA256_SHANI using the 'x86_shani(1way;2way)' SHA256 implementation`
|                3.92 |      255,199,693.76 |    0.1% |      0.04 | `SHA256_SSE4 using the 'standard;sse41(4way)' SHA256 implementation`
|                3.93 |      254,653,798.16 |    0.3% |      0.04 | `SHA256_STANDARD using the 'standard' SHA256 implementation`

fanquake commented:
What is the status here? This was recently re-opened, but seems to go in the opposite direction of your other PRs (by adding more MSVC specific code), such as #31507 (where you are dropping support for MSVC, and related workarounds).

hebasto commented:
> This was recently re-opened...

Yes. It was re-opened in response to @theuni's offline note regarding a potential loss of compiler diversity after 96d2380dc40e703e342dc0fad5d294cc49b25ff9.

> ... but seems to go in the opposite direction of your other PRs (by adding more MSVC specific code), such as #31507 (where you are dropping support for MSVC, and related workarounds).

It contradicts only a single [commit](https://github.com/bitcoin/bitcoin/commit/96d2380dc40e703e342dc0fad5d294cc49b25ff9) in #31507, which drops support for MSVC.

Hopefully, we can clarify the direction going forward during today's weekly IRC meeting.

maflcko commented:
> Hopefully, we can clarify the direction going forward during today's weekly IRC meeting.

The outcome was to close this?

fanquake commented:
I think so, and it seems like the plan is to go ahead with #31507?
